### ISSUE TITLE
Solver not working when game clearly on screen and visible

### ISSUE THREAD

**Issue by flexagontnt** [gtw123/OpusSolver#3]
When I press Win+Shift+A during the game, it shows an error message that says, "Please ensure that the Opus Magnum Window is completely visible and is on screen", despite the fact that my game is visible and on screen. How do I resolve this? Also, the lightning.exe does not work.

**Comment by gtw123** [comment]
If you have any overlays visible (like an FPS counter), try disabling them and see if it helps. If not, attach Opus.log to this issue and I'll see if I can figure out what's going wrong.

Can you clarify what you mean when you say that lightning.exe does not work?

**Comment by flexagontnt** [comment]
What is Opus.log? And the lightning.exe just does not open. Can you also post a guide on how to download and use the solver? I don't think I have any overlays. Here is the log:

**2019-02-06 17:21:16,194 INFO  Opus.ProgramMain: Starting up
2019-02-06 17:21:29,322 INFO  Opus.UI.Analysis.ScreenAnalyzer: Finding Opus Magnum process
2019-02-06 17:21:29,344 INFO  Opus.UI.Analysis.ScreenAnalyzer: Checking if window is completely on the screen
2019-02-06 17:21:29,352 INFO  Opus.WindowUtils: GetWindowRect: {Left = -3, Top = -26, Right = 1369, Bottom = 771}
2019-02-06 17:21:29,359 WARN  Opus.WindowUtils: GetWindowTextLength: error 1300
2019-02-06 17:21:29,359 INFO  Opus.WindowUtils: Found overlapping window "": {Left = 0, Top = 0, Right = 1, Bottom = 1}
2019-02-06 17:21:29,359 INFO  Opus.WindowUtils: Another window overlaps the window
2019-02-06 17:21:29,450 ERROR Opus.ProgramMain: Opus.UI.Analysis.AnalysisException: Please ensure the Opus Magnum window is completely visible and on screen.
   at Opus.UI.Analysis.ScreenAnalyzer.GetGameWindowBounds() in C:\Users\Zack\Desktop\OpusSolver-master\Opus\UI\Analysis\ScreenAnalyzer.cs:line 25
   at Opus.UI.Analysis.ScreenAnalyzer..ctor() in C:\Users\Zack\Desktop\OpusSolver-master\Opus\UI\Analysis\ScreenAnalyzer.cs:line 15
   at Opus.Form1.CreateAnalyzer() in C:\Users\Zack\Desktop\OpusSolver-master\Opus\Form1.cs:line 42**

**Comment by gtw123** [comment]
Lightning.exe is the name of the Opus Magnum exe - I don't know why it's called that!

Looking at the log, it thinks there's a 1x1 pixel window on top of the game window. Unfortunately that 1x1 window has no name so it's hard to tell what it might be.

A couple of things you could try:
* Try running the game in windowed mode (or if you're already doing that, try in fullscreen).
* You could disable the code that checks for overlapping windows: comment out line 25 of Opus/UI/Analysis/ScreenAnalyzer.cs

I thought about making a proper built release of the solver that people can download but I don't want to make it too simple for people to use it to "cheat" at the game. Also, it would be pretty difficult to get it work reliably on every computer as there a lot of things that can wrong when it tries to find the Opus Magnum window and I don't really have the resources to test it on lots of different computers.

**Comment by flexagontnt** [comment]
Now, when press Win+Shift+A when on an empty puzzle, I get a different type of error, 
![image](https://user-images.githubusercontent.com/47375581/52451497-125d4500-2af4-11e9-9944-53102b8328e2.png)
Can you help me please?

**Comment by gtw123** [comment]
It looks like it got as far as analysing the left hand side of the game screen, but it got confused and tried to analyse an invalid part of the screen. Could you post Opus.log again? If it's a bit big, you can attach it to this issue rather than copy/pasting it.

**Comment by flexagontnt** [comment]
This is from the most recent time I used it, which was yesterday, 

**2019-02-07 16:14:50,631 INFO  Opus.UI.Analysis.ScreenAnalyzer: Finding Opus Magnum process
2019-02-07 16:14:50,639 INFO  Opus.UI.Analysis.ScreenAnalyzer: Checking if window is completely on the screen
2019-02-07 16:14:50,639 INFO  Opus.WindowUtils: GetWindowRect: {Left = -3, Top = -26, Right = 1369, Bottom = 771}
2019-02-07 16:14:50,639 INFO  Opus.WindowUtils: GetClientRect: {Left = 0, Top = 0, Right = 1366, Bottom = 768}
2019-02-07 16:14:50,639 INFO  Opus.WindowUtils: ClientToScreen(topLeft): 0, 0
2019-02-07 16:14:50,639 INFO  Opus.WindowUtils: ClientToScreen(bottomRight): 1366, 768
2019-02-07 16:14:50,639 INFO  Opus.UI.Analysis.ScreenAnalyzer: Finding window bounds
2019-02-07 16:14:50,639 INFO  Opus.WindowUtils: GetClientRect: {Left = 0, Top = 0, Right = 1366, Bottom = 768}
2019-02-07 16:14:50,639 INFO  Opus.WindowUtils: ClientToScreen(topLeft): 0, 0
2019-02-07 16:14:50,639 INFO  Opus.WindowUtils: ClientToScreen(bottomRight): 1366, 768
2019-02-07 16:14:50,639 INFO  Opus.UI.Analysis.ScreenAnalyzer: Bounds: {X=0,Y=0,Width=1366,Height=768}
2019-02-07 16:14:50,639 INFO  Opus.UI.Analysis.ScreenAnalyzer: Analyzing game window
2019-02-07 16:14:50,698 INFO  Opus.UI.Analysis.HexGridAnalyzer: Finding center hex. Starting location: {X=843,Y=234}
2019-02-07 16:14:50,698 INFO  Opus.UI.Analysis.HexGridAnalyzer: Looking for vertical edge of hex in {X=761,Y=138,Width=164,Height=192}
2019-02-07 16:14:50,716 INFO  Opus.UI.Analysis.HexGridAnalyzer: Approximate center of hex: {X=806,Y=222}
2019-02-07 16:14:50,716 INFO  Opus.UI.Analysis.SidebarAnalyzer: Scrolling sidebar to top
2019-02-07 16:14:50,716 INFO  Opus.MouseUtils: Dragging from {X=0,Y=0} to {X=0,Y=767}; startFlags: RightDown; endFlags: RightUp
2019-02-07 16:14:50,862 INFO  Opus.MouseUtils: Dragging from {X=0,Y=0} to {X=0,Y=1}; startFlags: RightDown; endFlags: RightUp
2019-02-07 16:14:50,982 INFO  Opus.UI.Analysis.SidebarAnalyzer: Analyzing sidebar
2019-02-07 16:14:50,994 INFO  Opus.UI.Analysis.SidebarAnalyzer: Sidebar screen rect: {X=32,Y=0,Width=250,Height=498}
2019-02-07 16:14:50,995 INFO  Opus.UI.Analysis.SidebarAnalyzer: Sidebar is scrollable: True
2019-02-07 16:14:51,025 INFO  Opus.MouseUtils: Dragging from {X=32,Y=1} to {X=32,Y=0}; startFlags: RightDown; endFlags: RightUp
2019-02-07 16:14:51,145 INFO  Opus.MouseUtils: Dragging from {X=32,Y=0} to {X=32,Y=1}; startFlags: RightDown; endFlags: RightUp
2019-02-07 16:14:51,260 INFO  Opus.UI.Analysis.SidebarAnalyzer: Sidebar is continuous scrolling: True
2019-02-07 16:14:51,260 INFO  Opus.UI.Analysis.SidebarAnalyzer: Finding palettes
2019-02-07 16:14:51,260 INFO  Opus.UI.Analysis.ContinuousScrollingPaletteFinder: Analyzing continuously scrolling palettes
2019-02-07 16:14:51,260 INFO  Opus.UI.Analysis.ContinuousScrollingPaletteFinder: Capturing section 0 of the sidebar
2019-02-07 16:14:51,278 INFO  Opus.MouseUtils: Dragging from {X=32,Y=551} to {X=32,Y=0}; startFlags: RightDown; endFlags: RightUp
2019-02-07 16:14:51,411 INFO  Opus.UI.Analysis.ContinuousScrollingPaletteFinder: Capturing section 1 of the sidebar
2019-02-07 16:14:51,454 ERROR Opus.ProgramMain: System.OutOfMemoryException: Out of memory.
   at System.Drawing.Bitmap.Clone(Rectangle rect, PixelFormat format)
   at Opus.ScreenCapture.Clone(Rectangle rect, Boolean save) in C:\Users\Zack\Desktop\OpusSolver-master\Opus\Utils\ScreenCapture.cs:line 79
   at Opus.UI.Analysis.ContinuousScrollingPaletteFinder.GetSidebarCaptures() in C:\Users\Zack\Desktop\OpusSolver-master\Opus\UI\Analysis\ContinuousScrollingPaletteFinder.cs:line 76
   at Opus.UI.Analysis.ContinuousScrollingPaletteFinder.CaptureWholeSidebar() in C:\Users\Zack\Desktop\OpusSolver-master\Opus\UI\Analysis\ContinuousScrollingPaletteFinder.cs:line 34
   at Opus.UI.Analysis.ContinuousScrollingPaletteFinder.Analyze() in C:\Users\Zack\Desktop\OpusSolver-master\Opus\UI\Analysis\ContinuousScrollingPaletteFinder.cs:line 25
   at Opus.UI.Analysis.SidebarAnalyzer.Analyze() in C:\Users\Zack\Desktop\OpusSolver-master\Opus\UI\Analysis\SidebarAnalyzer.cs:line 34
   at Opus.UI.Analysis.ScreenAnalyzer.Analyze() in C:\Users\Zack\Desktop\OpusSolver-master\Opus\UI\Analysis\ScreenAnalyzer.cs:line 53
   at Opus.Form1.SolvePuzzle(ScreenAnalyzer analyzer) in C:\Users\Zack\Desktop\OpusSolver-master\Opus\Form1.cs:line 60
**

**Comment by gtw123** [comment]
Thanks for that! It turns out there was a bug which only happens on small screens. I think I've fixed it now. I've also added a workaround that should hopefully fix the problem it was having earlier with that invisible 1x1 window.

Can you get the latest version and try again? (If you cloned it with Git, do "git pull". If you downloaded the source as a zip file, download it again. Then build it again.)

If that gets a bit further then has another error, you may also need to edit App.config and set `MouseUtils.GlobalDragDelay` to 50, then build again.

**Comment by flexagontnt** [comment]
What is the recommended resolution for the game for the bot?

**Comment by gtw123** [comment]
I've mainly tested it at 2560x1440 and 1920x1080, but it does seem to work at lower resolutions too (at least on my computer).

The game itself is pretty hard to play at anything below 1920x1080 though.

**Comment by flexagontnt** [comment]
My game is set to 1366x768 for the bot. Will any problems occur?

**Comment by flexagontnt** [comment]
Actually, the bot does not work with a resolution **higher** than 1366x768 on my computer.

**Comment by gtw123** [comment]
What is the native resolution of your screen?

If you try to run the game at higher than 1366x768, what error does the bot give? Can you attach the log when that error happens?

**Comment by flexagontnt** [comment]
I think my native resolution is 1366x768. I think the error it gives is that it can't identify a glyph.

**Comment by gtw123** [comment]
Opus Magnum lets you choose a higher resolution than your screen normally allows. For example, if your screen is set to 1366 x 768 in the system control panel, the Options menu in Opus Magnum will actually let you select something like 1600 x 900. When you do this, it scales all the graphics down a bit to "simulate" a higher resolution screen. But since your screen is actually using a lower resolution, this will introduce artifacts and blurriness on the screen (especially for the in-game text). It also breaks the bot, which assumes the game is not scaling the graphics at all.

In other words, the bot requires the game to be running at the same resolution at what's selected in the system control panel. So if you can't set your screen higher than 1366 x 768 in the system control panel, I'm afraid that's the best you'll be able to use for the bot.

Anyway, it sounds like it's working OK for you now at 1366 x 768 so I'll close this issue now.
